python/Ganga/GPIDev/Credentials2/CredentialStore.py

```diff
@@ -0,0 +1,259 @@
+        If ``default`` is not given, it defaults to ``None``, so that this method never raises a ``KeyError``.
+
+        Args:
+            query (ICredentialRequirement):
+            default (ICredentialInfo):
+
+        Returns:
+            A single ICredentialInfo object which matches the requirements or ``default``
+        """
+        try:
+            return self[query]
+        except KeyError:
+            return default
+
+    def get_all_matching_type(self, query):
+        # type: (ICredentialRequirement) -> Sequence[ICredentialInfo]
+        """
+        Returns all ``ICredentialInfo`` with the type that matches the query
+
+        Args:
+            query (ICredentialRequirement):
+
+        Returns:
+            list[ICredentialInfo]: An list of all matching objects
+        """
+
+        return [cred for cred in self.credentials if isinstance(cred, query.info_class)]
+
+    def matches(self, query):
+        # type: (ICredentialRequirement) -> Sequence[ICredentialInfo]
+        """
+        Search the credentials in the store for all matches. They must match every condition exactly.
+
+        Args:
+            query (ICredentialRequirement):
+
+        Returns:
+            list[ICredentialInfo]: An list of all matching objects
+        """
+
+        return [cred for cred in self.get_all_matching_type(query) if cred.check_requirements(query)]
+
+    def match(self, query):
+        # type: (ICredentialRequirement) -> ICredentialInfo
+        """
+        Returns a single match from the store
+
+        Args:
+            query (ICredentialRequirement):
+
+        Returns:
+            ICredentialInfo: A single credential object. If more than one is found, the first is returned
+        """
+
+        matches = self.matches(query)
+        if len(matches) == 1:
+            return matches[0]
+        if len(matches) > 1:
+            logger.debug('More than one match...')
+            # If we have a specific object and a general one. Then we ask for a general one, what should we do.
+            # Does it matter since they've only asked for a general proxy? What are the use cases?
+            return matches[0]  # TODO For now just return the first one... Though perhaps we should merge them or something?
+        return None
+
+    @export
+    def renew(self):
+        # type: () -> None
+        """
+        Renew all credentials which are invalid or will expire soon.
+        It also uses the entries in `needed_credentials` and adds and renews those
+        TODO Should this function be standalone?
+        """
+        for cred in self.credentials:
+            if not cred.is_valid() or cred.time_left() < timedelta(hours=1):
+                cred.renew()
+        for cred_req in needed_credentials - self.credentials:
+            try:
+                self[cred_req].renew()
+            except KeyError:
+                self.create(cred_req)
+
+    @export
+    def clear(self):
```

Comment:
Should this call some destroy method on credentials within the system rather than just forgetting about them?
I don't know the pros/cons of this but it just seems like a sensible question that I don't know the answer to,